### ISSUE TITLE
Checks , Questions to T:Wayland and T:X11

### ISSUE THREAD

**Issue by RalfWerner** [suhan-paradkar/termux-wayland#8]
Thanks for proceeding X11 development. Example video with Floating Termux in XServer, swapped and split

https://user-images.githubusercontent.com/45426380/125047883-b241a300-e09f-11eb-9952-09bd1f22f905.mp4

<details><summary> What to see here in video and what not </summary>

The video shows typical floating and split representations of the Apps _Termux_ and _T:*_ for a device that allows and contains comments on the actions. The XClient (`touch`) shows the background (`*.jpg`), a 1st window with BT mouse (Blue Tooth is not treated here) and a 2nd window with the `*.jpg` in the app screensize + _rotated_ in black. At startup is _T:*_ fullscreen with Termux _in Picture_ and at the end 50% split from both apps on top and on-screen keyboard (OS-KB) and both extra keys (only possible here).

Before the video (first picture), Termux (bash) was started on the Pad. In `.bashrc`, the start of the XServer is included. All XClients for Window, File and Application Manager has stopped or not started. `touch` So mainly processes the _ConfigureNotify_ Events, all of which are implemented correctly.

I find _**no** differences_ between _T:Wayland_ and _T:X11_ (both XServer) and also when managing the Extra Keys (EK-t/x) but these are not displayed at Floating OS-KB or XServer and _T:Wayland_ change the EK-t behavior - more below in prep.

I could also repeat the effects (restricted), on a device that supports only split and floating keyboard (e.g. Gboard of AVD). These devices are very limited to the "parallel use" of the apps very limited. That's why I use _**toggle** functions_ in this case to switch between the apps (here `X11`) but also for keyboard (on/off), Multi Sessions and Window Manager (WM)

The EK-t is always displayed in the PiP/Split if you are not suppressed with _long tab_ in the drawer even _without_ OS-KB. The distribution is adjusted to the Termux width. I use here _one_ line with _popup_ extension.
If you want to repeat this use `fluxbox` instead of `touch` and write in` .termux/termux.properties`:
```
extra-keys-style=arrows-all
shortcut.next-session=ctrl+n
hide-soft-keyboard-on-startup=true
soft-keyboard-toggle-behaviour=enable/disable
#soft-keyboard-toggle-behaviour=show/hide
enforce-char-based-input=true
extra-keys=[[{macro:CTRL,display:'^',popup:{macro:'CTRL d',display:d}},TAB,{key:LEFT,popup:BKSP},{key:RIGHT,popup:{macro:'o ENTER',display:off}},{key: UP  ,popup:PGUP},{key:DOWN,popup:PGDN},{key:ENTER,popup:{macro:'CTRL n',display:'next'}},{key:KEYBOARD,popup:DRAWER},{macro:'mc SPACE i ENTER'  ,display:mc,popup:F10},{display:X11,macro:'t SPACE 0 ENTER',popup:{macro:'ob ENTER',display:OB}}]]
```
To start _T:*_ and toggle - when not already done - extend `.bashrc` by:
```
ct="am start -n com.termux" y="$ct.x11/.MainActivity" z="$ct/.app.TermuxActivity"; alias t="bash .bashrc"
if [ "$1" = 0 ];then $y;exit; elif [ "$1" = 2 ];then am $t; exit; fi ;X=Xwayland; px=$(pgrep -o $X); typeset -i i=0
p=$PREFIX/lib;t=$PREFIX/tmp;export LD_PRELOAD=$p/libtermux-exec.so:$p/libandroid-shmem.so XDG_RUNTIME_DIR=$t
while [ -z "$px" -a $i -lt 2 ];do $y;sleep 1;$X :0>$t/TX 2>&1&;$z; echo Start:$X try $i; px=$(pgrep -o $X);i+=1; done
```
</details>
<details><summary> *.apk/app management of Termux and T:X11(Wayland) plugins</summary>

Acceptance: Termux app and all plugins (_*.apk_) were created all with dev-keystroke. 
I would have preferred an XServer as a package in repo Termux-x11, like `sshd` from `openssh` package but @twaik could not convince @xeffyr (1st approach) and finally had fun on services. The advantage of the package approach is that Termux had direct influence on the Xwayland process (`$px` above). Currently, both apps must be managed in parallel and partially restarted.

In addition to installing _*.apk_ files (from you, @termux or self-made) with the Android App Installer or _Drag-Drop_ in AVD, I use studio and install directly on device (one $arch). For my _self-made *.apk_ I could use CompileSdk=TargetSdk=30 so together with [Boot-Loop](https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-780625381) the target **29+** to be able to get back Google playstore compatiblity, could simply reached.

Why did you _**rename X11 in Wayland?**_ Fortunately, the start is still `$y` (`*.x11` above) and up to your change of the version, the various _*.apk_ could be easily replaced. Now uninstall is required when it should be reset to version 1
With [dev-X11](https://www.dropbox.com/s/n85m6zkxkcvrk9p/dev-x11.apk?dl=0) I have the last version of @twaik with the dev-keystroke expansion of @xeffyr, which is about half as large (better/higher zip) like yours and my (about 7MB).
</details>
<details><summary>Problems with On-Screen-Keyboard (OS-KB) and Extra keys (EK) </summary>

Maybe I did not use _**T:Wayland correctly?**_ I could not discover any changes in XServer behaviour and EK-X has remained unchanged. Should not _**other keys appear there?**_ Can you explain to me how I can _**expand the keys?**_

My request at @agnostic-apollo to merge The two EK, he has rejected (says X11 is suspended). Could you _**check it and possibly do?**_ Both EK have PRO and CONS, but are only needed _once._ "Switch on" at EK-x is in Preferences and EK-t Long Tap in the Drawer.
My concerns with EK-Termux are only partially solved although @agnostic-apollo has complemented a lot of code in recent months also for harmonizing Termux and other plugins but **without X11** (suspended). With _T:Wayland_ add more bugs. E.g for me the popup "Next" Toggle (above) is no longer working after installation.
The OS-KB always appears once. But it is unclear to which app it belongs. This also applies to BT/USB KB. The OS-KB Toggle at _T:*_ is done with the back key, but the display stops in Termux (Switch to start screen). The app and all session remain active and _termux_ will be in other actions e.g. About _T:Widget_ again displayed with recent and another session.

### Great would be:
_one_ Extra Key block with `*termux.properties` EK description (above), which is displayed for _both_ apps, if not "switch off", it only needs _one_ OS-KB-toggle-button, the floating OS-KB How treated like the fixed and only toggled with this key. The active app for KB inputs could be changed with tap on the picture or split without disappears OS-KB from the screen. I am always surprised by reality. 
_**What are your experiences with this topic on your PiP device?**_

My API and Java knowledge is limited and I'm too old as a pensioner to do that effectively, but can support. I trust myself to understand code or generate an XClient in `C` which a window is manage, corresponds to the _EK-x_: 1-2 lines with horizontal draw/sweep at the bottom of the screens. Unfortunately, this does not helps anything in Android.
</details>

**Comment by suhan-paradkar** [comment]
@RalfWerner So.. I assume that you need those features implemented in T: Wayland

**Comment by RalfWerner** [comment]
> Okay.. thats fine... (to ... please wait)

Finished (first [commit](https://github.com/suhan-paradkar/termux-wayland/issues/8#issue-940543643) here)!

**Comment by suhan-paradkar** [comment]
> I find no differences between T:Wayland and T:X11 (both XServer) and also when managing the Extra Keys (EK-t/x)

That's because as of now.. only bug fixes have been done to this fork....

> For my self-made *.apk I could use CompileSdk=TargetSdk=30 so together with Boot-Loop the target 29+ to be able to get back Google playstore compatiblity, could simply reached.

On bumping SDK version.... I have been giving ut a lot if thought.. since only termux app has problems with it... Not T: Wayland....
But then... SharedUid would not be used.. since it is now deprecated... So.. a whole new implementation has to prepared from scratch...

> Problems with On-Screen-Keyboard (OS-KB) and Extra keys (EK)

Well thats clearly a work in progress currently...
Currently the EK artifacts are experiencing unexpected crashes.. also that may be due to the fact that the support libraries have been replaced by androidx libraries in that branch and I need logs regarding the same... ( I can't get them since I don't own a PC)

> one Extra Key block with *termux.properties EK description (above), which is displayed for both apps, if not "switch off"

Well.. I could implement that... But I require some time.....

**Comment by suhan-paradkar** [comment]
> The EK-t is always displayed in the PiP/Split if you are not suppressed with long tab in the drawer even without OS-KB

Currently the EK implementation is hardcoded and isn't changed...

**Comment by RalfWerner** [comment]
Thank you for the answers - some of my 6 _**?-questions**_ above you've clarified! I wish you much success in your work on the code.
> On Bumping SDK Version ...  a whole new implementation has to prepared from scratch

@agnostic-apollo has already done a Termux-general-revision between v108 and v109. He is the best expert for Android implementation. The current v117 has twice as many sources such as the _Android 10_ branch (v98), which is hardly useful, rather the _updated flover_ branch of @fornwall (v112) and the SDK Bump is thus proven - it works.

I think the _boot loop_ integration in the _rolling_ termux app is with the adaptation of two sources: **1.** `*/app/build.gradle` and **2.** `*/TermuxInstaller.java` possible without hindering the current corrections or to affect the build. To do the **EK** Merge you should have to convince @agnostic-apollo to insert
`manifestPlaceholders.TERMUX_X11_APP_NAME = "Termux:X11 "` behind [1.](https://github.com/termux/termux-app/blob/master/app/build.gradle#L39) and in addition to support @termux.

XServer Integration in Termux is a bit more complex, but works well in my checks. Currently the _T:X11_ plugin is necessary and a few small extensions in the _bootstrap_.

You speak about _crashes_ and _bugs_ - I've none - which are _that and how **can I repeat them with T:X11?**_

**Comment by Heasterian** [comment]
Hi!

I have two questions.

First: Did app comes with same signature as Termux from fdroid so It can be used flawlessly with it?

Second: Would it be possible to fix mouse movement in touchpad mode to work with games rotating camera using it's movement (ie. Amnesia: The Dark Descent in box64)?

**Comment by suhan-paradkar** [comment]
> @agnostic-apollo has already done a Termux-general-revision between v108 and v109. He is the best expert for Android implementation.

Yes He is.... I would also like him to help.. since currently only I am working on this...

> The current v117 has twice as many sources such as the _Android 10_ branch (v98), which is hardly useful, rather the _updated flover_ branch of @fornwall (v112) and the SDK Bump is thus proven - it works.

I will be checking that out...

> 
> I think the _boot loop_ integration in the _rolling_ termux app is with the adaptation of two sources: **1.** `*/app/build.gradle` and **2.** `*/TermuxInstaller.java` possible without hindering the current corrections or to affect the build. To do the **EK** Merge you should have to convince @agnostic-apollo to insert
> `manifestPlaceholders.TERMUX_X11_APP_NAME = "Termux:X11 "` behind [1.](https://github.com/termux/termux-app/blob/master/app/build.gradle#L39) and in addition to support @termux.

Yes...

> XServer Integration in Termux is a bit more complex, but works well in my checks. Currently the _T:X11_ plugin is necessary and a few small extensions in the _bootstrap_.
> 
> You speak about _crashes_ and _bugs_ - I've none - which are _that and how **can I repeat them with T:X11?**_

Those are only repeatable on the build artifacts of #4 . Also.. as some users complain... On using direct mode in T: Wayland....in which it supposedly takes place after some time....

**Comment by suhan-paradkar** [comment]
> Hi!
> 
> I have two questions.
> 
> First: Did app comes with same signature as Termux from fdroid so It can be used flawlessly with it?

As of now... No.. since @termux has archived the repo

> Second: Would it be possible to fix mouse movement in touchpad mode to work with games rotating camera using it's movement (ie. Amnesia: The Dark Descent in box64)?

Yes.. but only after #4 is merged...

**Comment by agnostic-apollo** [comment]
>Yes He is.... I would also like him to help.. since currently only I am working on this...

Currently, unfortunately, I can't pick up another project, already working on other stuff, and `termux-app` already requires lot of fixes/updates. I may look into this in future. However, if relatively small changes  need to be made in `termux-app`, I should be able to help with that or help out if you get stuck and need some verbal help in `X11`. I haven't used the `X11` app myself  so don't know exactly what needs to be changed in `termux-app` to improve the experience, including what changes are required in extra keys. @suhan-paradkar I need well explained details of the problem and suggested solutions.

>have to convince @agnostic-apollo to insert
manifestPlaceholders.TERMUX_X11_APP_NAME = "Termux:X11 " behind 1

@RalfWerner asked me to add `X11` info in `TermuxConstants` earlier but I didn't cause there was no internal usage in `termux-app` for it. The `manifestPlaceholders` are **only** variables that can be used in [`AndroidManifest.xml`](https://github.com/termux/termux-app/blob/v0.117/app/src/main/AndroidManifest.xml), like done for `sharedUserId` instead of hard coding values like `com.termux`. The `manifestPlaceholders` for plugin apps aren't even being used in `termux-app` `AndroidManifest.xml`, they are only left as reference for how to use them in `AndroidManifest.xml` of plugin apps, when in future hardcoded values are removed from them. The same thing is done in [`strings.xml`](https://github.com/termux/termux-app/blob/v0.117/app/src/main/res/values/strings.xml) with `ENTITY` variables and [`TermuxContants.java`](https://github.com/termux/termux-app/blob/v0.117/termux-shared/src/main/java/com/termux/shared/termux/TermuxConstants.java) class. Their **only** goal is that hard coded values are not used, specially so that if someone forks `termux-app`, only limited changes in these places need to be made instead of everywhere. They do not affect functionality of other plugin apps in any way.

As for why `X11` isn't officially supported and was never released, likely because it was never stable but I don't know for sure, it was before my time and I have never used or looked into it. Ask xeffyr or grimler for such details and if it can be released after you guys have fixed the bugs.

>since only termux app has problems with it... Not T: Wayland....
But then... SharedUid would not be used.. since it is now deprecated... So.. a whole new implementation has to prepared from scratch...

The way `targetSdkVersion` works with `sharedUserId` is that the first app that's installed, its `targetSdkVersion` is used. So if you bump it to `29` for a plugin and user installs that plugin before `termux-app`, then `termux-app` will not work on android 10+. This was from my own tests on AVD and may be different on real devices. Basically, either all termux apps should be published to play store together or not at all. Yes, `sharedUserId` has been deprecated but I haven't **seen** any commits in AOSP since deprecation announcement of it being removed internally in `AOSP`. It's still being used a lot in `AOSP`. And if it is removed, then `termux-app` APK system will be gone too and playstore releases won't be doable anyways, unless an alternative is provided officially with similar functionality, or a system is created in `termux-app` that complies with playstore polices.

**Comment by agnostic-apollo** [comment]
If you plan on releasing on F-Droid do not use github packages as per 22349a7d. Dependencies from private repositories cannot be imported by apps as per F-Droid policy. Import `termux-shared` from Jitpack as per [`v0.116`](https://github.com/termux/termux-app/releases/tag/v0.116) release notes. Also if possible, remove hard coded `com.termux` values if any exist in `X11`.

**Comment by suhan-paradkar** [comment]
> Currently, unfortunately, I can't pick up another project, already working on other stuff, and `termux-app` already requires lot of fixes/updates. I may look into this in future. 

Thats fine...

>However, if relatively small changes need to be made in `termux-app`, I should be able to help with that or help out if you get stuck and need some verbal help in `X11`. I haven't used the `X11` app myself so don't know exactly what needs to be changed in `termux-app` to improve the experience, including what changes are required in extra keys. @suhan-paradkar I need well explained details of the problem and suggested solutions.

Currently Extra-keys are WIP in this fork... You can see #4 for the details..
As per now... I don't think any changes need yo be done to Termux-app.... But I don't guarantee that tho...

> @RalfWerner asked me to add `X11` info in `TermuxConstants` earlier but I didn't cause there was no internal usage in `termux-app` for it. The `manifestPlaceholders` are **only** variables that can be used in [`AndroidManifest.xml`](https://github.com/termux/termux-app/blob/v0.117/app/src/main/AndroidManifest.xml), like done for `sharedUserId` instead of hard coding values like `com.termux`. The `manifestPlaceholders` for plugin apps aren't even being used in `termux-app` `AndroidManifest.xml`, they are only left as reference for how to use them in `AndroidManifest.xml` of plugin apps, when in future hardcoded values are removed from them. The same thing is done in [`strings.xml`](https://github.com/termux/termux-app/blob/v0.117/app/src/main/res/values/strings.xml) with `ENTITY` variables and [`TermuxContants.java`](https://github.com/termux/termux-app/blob/v0.117/termux-shared/src/main/java/com/termux/shared/termux/TermuxConstants.java) class. Their **only** goal is that hard coded values are not used, specially so that if someone forks `termux-app`, only limited changes in these places need to be made instead of everywhere. They do not affect functionality of other plugin apps in any way.
> 
> As for why `X11` isn't officially supported and was never released, likely because it was never stable but I don't know for sure, it was before my time and I have never used or looked into it. Ask xeffyr or grimler for such details and if it can be released after you guys have fixed the bugs.
> 
> > since only termux app has problems with it... Not T: Wayland....
> > But then... SharedUid would not be used.. since it is now deprecated... So.. a whole new implementation has to prepared from scratch...
> 
> The way `targetSdkVersion` works with `sharedUserId` is that the first app that's installed, its `targetSdkVersion` is used. So if you bump it to `29` for a plugin and user installs that plugin before `termux-app`, then `termux-app` will not work on android 10+. This was from my own tests on AVD and may be different on real devices. Basically, either all termux apps should be published to play store together or not at all. Yes, `sharedUserId` has been deprecated but I haven't **seen** any commits in AOSP since deprecation announcement of it being removed internally in `AOSP`. It's still being used a lot in `AOSP`. And if it is removed, then `termux-app` APK system will be gone too and playstore releases won't be doable anyways, unless an alternative is provided officially with similar functionality, or a system is created in `termux-app` that complies with playstore polices.

**Comment by suhan-paradkar** [comment]
> @suhan-paradkar If you plan on releasing on F-Droid do not use github packages as per [22349a7](https://github.com/suhan-paradkar/termux-wayland/commit/22349a7d7d56f2dbdb2c383058c3f95b206f835b). Dependencies from private repositories cannot be imported by apps as per F-Droid policy. Import `termux-shared` from Jitpack as per [`v0.116`](https://github.com/termux/termux-app/releases/tag/v0.116) release notes. Also if possible, remove hard coded `com.termux` values if any exist in `X11`.

Yea.. that has to be done....
My plan is to remove all deprecated features and then implement it freshly....
After that I am planning to ask to release it to fdroid

**Comment by agnostic-apollo** [comment]
Okay, Goodluck.

**Comment by RalfWerner** [comment]
> @agnostic-apollo The `manifestPlaceholders` are only variable ... isn't officially supported and was never released, likely because it was never stable but I don't know for sure, it was before my time and I have never used or looked into it

I know, but it's a first step out of the _suspend_ mode, which was only set after the IDF appeal of @twaik (I hope he come back). I have accompanied the development from the beginning and work since this tilme with _T: X11_. For a periode Termux:X11 was released on playstore!
[Compared](https://github.com/RalfWerner/integrated-process/issues/2#issuecomment-594033201) to other XServers, this is the _most stable._ The Direct mode can become unstable (I'm just checking) But just because _T: X11_ is the _**only**_ XServer that _**has**_ MultiTouch capability!
@Xeffyr has accompanied and supported the process as far as its capacity has permitted. With the integration, however, the status of the _Termux-app_ would change from "Terminal Emutator" to "Linux-OS", which almost all developers do _not_ want.

**Comment by suhan-paradkar** [comment]
> Did I understood correctly that you also had _**no**_ crashes/bugs with basic _**T:X11 and without `Direct mode`?**_

Yes.. there were absolutely no crashes at all(Except direct mode)....

> The use of multitouch requires XClients that support it and this includes a _device analysis_ and processing of `GenericEvent`.

Will see about that...

**Comment by Heasterian** [comment]
As for hardware acceleration, support of DRI3 would be nice as we could use vulkan driver directly. Old Termux:X11 is able to load Zink and Turnip with KGSL support so hardware acceleration is possible if you load OpenGL driver on proot side.

Panvk is too young to even try to use it with zink, but someone with Mali GPU could try it. At least usage of glxinfo.

**Comment by Heasterian** [comment]
Ability to load Vulkan would let us use DXVK in box86 (and box64 in future), that's why it would be nice.

**Comment by RalfWerner** [comment]
> Will see about that...

Author of the Xorg/Xinput functions (Multi Touch) is Peter Hutterer (Shot). I contacted him because of the crashes. But he has no time to deal with Android implementations. An example (`multi.c` uses `cairo`) From his blog I have used with mods as XClient to check and like `touch` in the video, which checks various other XServer functions. Start here with `t t pet*` (he describes thumb actions)
![grafik](https://user-images.githubusercontent.com/45426380/125432648-568cb9c2-7c60-493c-a995-3a5972a5c876.png)

**Comment by agnostic-apollo** [comment]
> I know, but it's a first step out of the suspend mode

At **this** point in time, I have zilch knowledge of `termux:x11` and its stability/problems so I can't in good conscience approve or disapprove if it should be publicly released. Ask one of the other @termux maintainers who would know of the issues or have used it, likely xeffyr or grimler. I am still willing to add changes to `termux-app` unless an objection is raised by others.

>the status of the Termux-app would change from "Terminal Emutator" to "Linux-OS", which almost all developers do not want.

I don't think that would be an issue, termux already provides proot distros and x11 repo.

**Comment by RalfWerner** [comment]
>I don't think that would be an _**issue**_, termux already provides proot distros and x11 repo.

No _issue_ but support work. @xeffyr prepare, among other things, much **XClients** in the _x11 repo_ and makes a great job that is actually too much for one person. As long as Multitouch is meaningless or desirable that events is intercepted by the VNC viewer or the _performace loss_ is not measured, _termux_ does not **need** its own **XServer**.

Especially the excessive use of _proot distro_ (further _performance loss_) and often associated Window Manager (up to Windows 95) shows the increasing meaning of GUI/XServer for termux _user_. For @termux debug, it is less important and also the PC XServer can be used instead, if not for debug five fingers are have to be moved across the screen.

**Needs** are rising with knowledge and experience! Therefore, I recommend you @agnostic-apollo: _use T:X11_ by yourself. 
@suhan-paradkar here is a dedicated young developer with potential and many of the _Discord kids_ also take a lot of support work away from @termux.

**Comment by agnostic-apollo** [comment]
>shows the increasing meaning of GUI/XServer for termux user.

If its so important, then termux users should set up a bounty for it so that it increases chances that some experienced X11 people can come in and try to implement and fix stuff, be it external people or any current @termux devs. The above issues are likely not small issues and would be very time consuming and require in-depth knowledge. Even more time would be needed for continued future support, so it shouldn't be expected that people would just come in and do all the work for weeks/months. Free in FOSS is for freedom, not for free beer. Either wait for people to do it themselves in their free time or put a bounty on it so it "increases chances" of faster contributions.
